File: 0170-two-sum-iii-data-structure-design/0170-two-sum-iii-data-structure-design.java
class TwoSum {
    
    Map<Integer,Boolean> set;

    public TwoSum() {
        set = new HashMap<>();
    }
    
    public void add(int number) {
        set.compute(number, (k,v)->v != null);
    }
    
    public boolean find(int value) {
        for(Integer i:set.keySet()) {
            if(i == value-i && set.get(i))
                return true;
            else if(i != value-i && set.containsKey(value-i))
                return true;
        }
        return false;
    }
}

/**
 * Your TwoSum object will be instantiated and called as such:
 * TwoSum obj = new TwoSum();
 * obj.add(number);
 * boolean param_2 = obj.find(value);
 */

File: 0170-two-sum-iii-data-structure-design/0170-two-sum-iii-data-structure-design.cpp
class TwoSum {
public:
    
    unordered_map<int, int>* map;
    TwoSum() {
        map = new unordered_map<int, int>();
    }
    
    void add(int number) {
        (*map)[number]++;
    }
    
    bool find(int value) {
        for(unordered_map<int, int>::iterator it = map->begin();it!=map->end();it++) {
            if (((long)it->first == (long)value - it->first) && it->second>1) {
                return true;
            } else if (((long)it->first != (long)value - it->first) && map->find((long)value - it->first) != map->end()) {
                return true;
            }
        }
        return false;
    }
};

/**
 * Your TwoSum object will be instantiated and called as such:
 * TwoSum* obj = new TwoSum();
 * obj->add(number);
 * bool param_2 = obj->find(value);
 */